locator: util: optimize describe_ring

bhalevy opened the pull request:
This change includes basic optimizations to
locator::describe_ring, mainly caching the per-endpoint information in an unordered_map instead of looking them up in every inner-loop.

This yields an improvement of 20% in cpu time.
With 45 nodes organized as 3 dcs, 3 racks per dc, 5 nodes per rack, 256 tokens per node, yielding 11520 ranges and 9 replicas per range, describe_ring took Before: 30 milliseconds (2.6 microseconds per range) After:  24 milliseconds (2.1 microseconds per range)

Add respective unit test for vnode keyspace
and for tablets.

Fixes #24887

* backport up to 2025.1 as describe_ring slowness was hit in the field with large clusters

avikivity commented on api/storage_service.cc:
```diff
@@ -534,11 +534,17 @@ void unset_view_builder(http_context& ctx, routes& r) {
-    co_return json::json_return_type(stream_range_as_array(co_await ss.local().describe_ring(keyspace), token_range_endpoints_to_json));
+    apilog.info("describe_ring: keyspace={}: starting", keyspace);
+    auto ret = stream_range_as_array(co_await ss.local().describe_ring(keyspace), token_range_endpoints_to_json);
+    apilog.info("describe_ring: keyspace={}: done", keyspace);
```
This is way too noisy. And unnecessary, just remove it.

bhalevy replied:
ok

avikivity commented on locator/util.cc:
```diff
@@ -98,35 +98,38 @@ get_range_to_address_map(locator::effective_replication_map_ptr erm) {
-            tr._endpoint_details.push_back(details);
+            auto it = host_infos.find(endpoint);
+            if (it == host_infos.end()) {
+                it = host_infos.emplace(endpoint, get_describe_ring_endpoint_info(endpoint, topology, gossiper)).first;
```
Why is get_describe_ring_endpoint_info() so slow?

bhalevy replied:
Do you ask why it needed the optimization, or is there an indication it is so slow with this change?

bhalevy replied:
It just did a map lookup for every member:
- `gossiper.get_address_map().get(endpoint)`
- `topology.get_datacenter(endpoint)`
- `topology.get_rack(endpoint)`
- `gossiper.get_rpc_address(endpoint)` (that has a nested map lookup: first for the endpoint_state_ptr then for the RPC_ADDRESS state)

bhalevy replied:
This summed up to 20% reduction in runtime when idle.
The reason to optimize it at all is that it was observed to take a long time (over 30 seconds) in the field during repair (that also runs in the streaming scheduling group) and presumingly under query load as well.

avikivity replied:
So we're optimizing from over 30 seconds to over 24 seconds?

It's a nice improvement, but does it doesn't materially change anything, it's still to long for human interaction.

What's the cpu usage of the thing? Maybe the problem was scheduling, not runtime (and then the solution is to provide a separate nested scheduling groups for the data plane and for the control plan under streaming).

avikivity replied:
Then I don't understand what we're doing here. 

If we're optimizing by 20%, then whatever problem caused it to go over 30 seconds could cause it to go over 36 seconds and we're back where we started.

avikivity commented on test/cluster/test_multidc.py:
```diff
@@ -464,3 +465,68 @@ async def test_restart_with_prefer_local(request: pytest.FixtureRequest, manager
     await manager.server_start(s_info.server_id)
+
+@pytest.mark.asyncio
+async def test_vnodes_describe_ring(request: pytest.FixtureRequest, manager: ManagerClient, build_mode) -> None:
```
Don't we already have a test for this?

We should be able to test these APIs by instantiating a topology object with arbitrary node configuration, instead of having to launch a cluster.

bhalevy commented on api/storage_service.cc:
```diff
@@ -534,11 +534,17 @@ void unset_view_builder(http_context& ctx, routes& r) {
+    apilog.info("describe_ring: keyspace={} table={}: starting", keyspace, table);
+    auto ret = stream_range_as_array(co_await ss.local().describe_ring_for_table(keyspace, table), token_range_endpoints_to_json);
+    apilog.info("describe_ring: keyspace={} table={}: done", keyspace, table);
+    co_return ret;
```
ok

bhalevy commented:
In [fdb1ae0](https://github.com/scylladb/scylladb/commit/fdb1ae047ca55705ded6072404bbfb592dcc687e):
- removed extra log messages
- replaced fancy cluster test with a c++ boost unit test in tablets_test.cc

avikivity commented:
> * replaced fancy cluster test with a c++ boost unit test in tablets_test.cc

We should do more of this, the full-stack testing is killing us.

yaronkaikov commented:
> @scylladb/releng - is CI for this stuck?

As indicate in https://github.com/scylladb/scylladb/pull/24889#issuecomment-3140369046 the CI failed 2 weeks ago, since then i don't see any new push or any update on this PR. so there is no reason for the CI to run again

bhalevy commented:
> @bhalevy - please re-run CI.

done

avikivity commented:
The patch is okay, but I disagree with backporting it. Optimizing describe_ring from 30ms to 24ms doesn't seem like something that anyone will notice, unless it has a bug.

bhalevy commented:
In [50abeb1](https://github.com/scylladb/scylladb/commit/50abeb12706d580a46f11047a232418067b91907):
- test: cluster: test_repair: add test_vnode_keyspace_describe_ring
  - verifies describe_ring validity
- optimize the get_natural_replicas path:
  - locator: effective_replication_map: get_natural_replicas: get is_vnode param
  - vnode_effective_replication_map: do_get_replicas: throw internal error if token not found in map
  - locator: util: construct_range_to_endpoint_map: pass is_vnode=true to get_natural_replicas

raphaelsc commented:
> The patch is okay, but I disagree with backporting it. Optimizing describe_ring from 30ms to 24ms doesn't seem like something that anyone will notice, unless it has a bug.

There was a hypothesis that the problem was caused by memory pressure. The optimization is welcome of course but maybe the root cause was some contiguous allocation somewhere with large clusters, making it stall?? We should review the path for potential sites.

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -479,12 +479,12 @@ host_id_vector_replica_set vnode_effective_replication_map::do_get_replicas(cons
-    return do_get_replicas(tok, false);
+host_id_vector_replica_set vnode_effective_replication_map::get_replicas(const token& tok, bool is_vnode) const {
+    return do_get_replicas(tok, is_vnode);
 }
```
What does is_vnode mean? That it's one of some node's tokens?

Ok, that's what the patch changelog says.

But why is it important? All the data paths will look for the partition key's token. I feel it clutters the interface and slows down the data path (by a tiny bit) for an unimportant bit in the control path.

avikivity replied:
btw, such distinctions are better in the method name. Parameters are good when a caller can call with many values, but when each call site calls with a constant, it means you're multiplexing different things in a single function.

bhalevy replied:
> What does is_vnode mean? That it's one of some node's tokens?

Yes.  From the vnode perspective it's the vnode's first token (inclusive)

> 
> Ok, that's what the patch changelog says.
> 
> But why is it important? All the data paths will look for the partition key's token. I feel it clutters the interface and slows down the data path (by a tiny bit) for an unimportant bit in the control path.

The replication map is indexed by the vnode tokens.
When an arbitrary token is looked up we revert to a binary search (std::lower_bound) in `sorted_tokens`.
Although the latter has O(log N) complexity which isn't too bad it doing random access in memory and trashes the dcache, and it is done on every range where in a large cluster we can have tens of thousands of those.

bhalevy replied:
> btw, such distinctions are better in the method name. Parameters are good when a caller can call with many values, but when each call site calls with a constant, it means you're multiplexing different things in a single function.

The function (`vnode_effective_replication_map::do_get_replicas`) is already multiplxed:
```
   const token& key_token = _rs->natural_endpoints_depend_on_token()
        ? (is_vnode ? tok : _tmptr->first_token(tok))
        : default_replication_map_key;
```

Another approach I contemplated about was to define `struct search_token` that would have a boolean class indicating `is_vnode` (maybe using a better term) that would be passed down to `do_get_replicas`.

avikivity replied:
Ok, if it's already there, of course we can use it, but the criticisms for the API apply.
